File: backend/search-api/internal/domain/trip.go
package domain

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

// Trip represents the external Trip DTO from trips-api
// This is the structure returned by trips-api GET /trips/:id endpoint
type Trip struct {
	ID primitive.ObjectID `json:"id" bson:"_id,omitempty"`

	// Driver and ownership
	DriverID int64 `json:"driver_id" bson:"driver_id"`

	// Location information (from trips-api with simple lat/lng format)
	Origin      TripLocation `json:"origin" bson:"origin"`
	Destination TripLocation `json:"destination" bson:"destination"`

	// Trip timing
	DepartureDatetime        time.Time `json:"departure_datetime" bson:"departure_datetime"`
	EstimatedArrivalDatetime time.Time `json:"estimated_arrival_datetime" bson:"estimated_arrival_datetime"`

	// Pricing and availability
	PricePerSeat   float64 `json:"price_per_seat" bson:"price_per_seat"`
	TotalSeats     int     `json:"total_seats" bson:"total_seats"`
	AvailableSeats int     `json:"available_seats" bson:"available_seats"`
	ReservedSeats  int     `json:"reserved_seats" bson:"reserved_seats"`

	// Optimistic locking for concurrency control
	AvailabilityVersion int `json:"availability_version" bson:"availability_version"`

	// Vehicle and preferences
	Car         Car         `json:"car" bson:"car"`
	Preferences Preferences `json:"preferences" bson:"preferences"`

	// Trip details
	Status      string `json:"status" bson:"status"` // draft, published, full, in_progress, completed, cancelled
	Description string `json:"description" bson:"description"`

	// Cancellation info (optional)
	CancelledAt        *time.Time `json:"cancelled_at,omitempty" bson:"cancelled_at,omitempty"`
	CancelledBy        *int64     `json:"cancelled_by,omitempty" bson:"cancelled_by,omitempty"`
	CancellationReason string     `json:"cancellation_reason,omitempty" bson:"cancellation_reason,omitempty"`

	// Timestamps
	CreatedAt time.Time `json:"created_at" bson:"created_at"`
	UpdatedAt time.Time `json:"updated_at" bson:"updated_at"`
}

// TripLocation represents location data from trips-api (with simple lat/lng coordinates)
type TripLocation struct {
	City        string            `json:"city"`
	Province    string            `json:"province"`
	Address     string            `json:"address"`
	Coordinates SimpleCoordinates `json:"coordinates"`
}

// SimpleCoordinates represents simple lat/lng coordinates from trips-api
type SimpleCoordinates struct {
	Lat float64 `json:"lat"`
	Lng float64 `json:"lng"`
}

// ToLocation converts TripLocation (from trips-api) to Location (for search-api storage)
func (tl *TripLocation) ToLocation() Location {
	return Location{
		City:        tl.City,
		Province:    tl.Province,
		Address:     tl.Address,
		Coordinates: NewGeoJSONPoint(tl.Coordinates.Lat, tl.Coordinates.Lng),
	}
}

// ToSearchTrip converts Trip DTO to SearchTrip with driver information
// Driver info must be fetched separately from users-api
// This method converts simple lat/lng coordinates to GeoJSON format required by MongoDB
func (t *Trip) ToSearchTrip(driver Driver) *SearchTrip {
	return &SearchTrip{
		TripID:                   t.ID.Hex(),
		DriverID:                 t.DriverID,
		Driver:                   driver,
		Origin:                   t.Origin.ToLocation(),
		Destination:              t.Destination.ToLocation(),
		DepartureDatetime:        t.DepartureDatetime,
		EstimatedArrivalDatetime: t.EstimatedArrivalDatetime,
		PricePerSeat:             t.PricePerSeat,
		TotalSeats:               t.TotalSeats,
		AvailableSeats:           t.AvailableSeats,
		Car:                      t.Car,
		Preferences:              t.Preferences,
		Status:                   t.Status,
		Description:              t.Description,
		SearchText:               buildSearchText(t, driver),
		CreatedAt:                t.CreatedAt,
		UpdatedAt:                t.UpdatedAt,
	}
}

// buildSearchText creates a concatenated text field for full-text search fallback
func buildSearchText(trip *Trip, driver Driver) string {
	return trip.Origin.City + " " +
		trip.Origin.Province + " " +
		trip.Destination.City + " " +
		trip.Destination.Province + " " +
		driver.Name + " " +
		trip.Car.Brand + " " +
		trip.Car.Model + " " +
		trip.Description
}

File: mobile/src/types/trip.ts
/**
 * Trip-related types for the CarPooling Trips API (Port 8002)
 * Database: MongoDB
 */

/**
 * Trip status values
 */
export type TripStatus =
  | 'draft' // Trip created but not published
  | 'published' // Trip visible and accepting reservations
  | 'full' // All seats reserved
  | 'in_progress' // Trip has started
  | 'completed' // Trip finished successfully
  | 'cancelled' // Trip cancelled

/**
 * Geographic coordinates (latitude and longitude)
 */
export interface Coordinates {
  lat: number // Latitude
  lng: number // Longitude
}

/**
 * Location with city, province, address and coordinates
 */
export interface Location {
  city: string
  province: string
  address: string
  coordinates: Coordinates
}

/**
 * Vehicle information
 */
export interface Car {
  brand: string
  model: string
  year: number
  color: string
  plate: string
}

/**
 * Trip preferences for passengers
 */
export interface Preferences {
  pets_allowed: boolean
  smoking_allowed: boolean
  music_allowed: boolean
}

/**
 * Complete trip object
 */
export interface Trip {
  id: string // MongoDB ObjectID as hex string
  driver_id: number // User ID from users-api
  origin: Location
  destination: Location
  departure_datetime: string // ISO 8601 datetime
  estimated_arrival_datetime: string // ISO 8601 datetime
  price_per_seat: number
  total_seats: number
  reserved_seats: number
  available_seats: number
  availability_version: number // For optimistic locking
  car: Car
  preferences: Preferences
  status: TripStatus
  description: string
  cancelled_at?: string // ISO 8601 datetime, only if status is 'cancelled'
  cancelled_by?: number // User ID who cancelled
  cancellation_reason?: string
  created_at: string // ISO 8601 datetime
  updated_at: string // ISO 8601 datetime
}

/**
 * Data for creating a new trip
 */
export interface CreateTripRequest {
  origin: Location
  destination: Location
  departure_datetime: string // ISO 8601 datetime (RFC3339)
  estimated_arrival_datetime: string // ISO 8601 datetime (RFC3339)
  price_per_seat: number // Must be >= 0
  total_seats: number // 1-8 seats
  car: Car
  preferences: Preferences
  description?: string
}

/**
 * Data for updating an existing trip (all fields optional)
 * Note: Cannot update if trip has reservations
 */
export interface UpdateTripRequest {
  origin?: Location
  destination?: Location
  departure_datetime?: string // ISO 8601 datetime
  estimated_arrival_datetime?: string // ISO 8601 datetime
  price_per_seat?: number
  total_seats?: number
  car?: Car
  preferences?: Preferences
  description?: string
}

/**
 * Cancel trip request
 */
export interface CancelTripRequest {
  reason: string
}

/**
 * Query filters for listing trips
 */
export interface TripFilters {
  driver_id?: number
  status?: TripStatus
  origin_city?: string
  destination_city?: string
  page?: number // Default: 1
  limit?: number // Default: 10, max: 100
}

/**
 * Paginated trip list response
 */
export interface TripListResponse {
  trips: Trip[]
  total: number
  page: number
  limit: number
}
